### PR TITLE
#86 – Add sigma decay

### PR DESCRIPTION
**Pull request by mcoot**
PR to add support for a sigma decay feature.

### Changes

- Implements a sigma decay feature per #86
- New configs:
    - `ENABLE_TRUESKILL_SIGMA_DECAY` – enables the feature (off by default)
    - `TRUESKILL_SIGMA_DECAY_DELTA` – configures how much to increase sigma by on each decay event (current default is 0.05)
    - `TRUESKILL_SIGMA_DECAY_GRACE_DAYS` - configures a grace period before decay starts (current default is 1 day)
    - `TRUESKILL_SIGMA_DECAY_JOB_SCHEDULED_TIME` – configures ISO format time for the time when the decay job should be executed (defaults to midnight UTC)
- Adds a new column to the `PlayerCategoryTrueskill` model for the last timestamp that a game was finished for that PCT, which gets set on game completion
- Adds the sigma decay job

### Tasks

- [x] Testing – see comments
- [ ] We should decide on the default values of the new parameters

### CODE REVIEW

**Comment by mcoot** [comment]
Not ready to go yet, probably should test further, but I did manage to do a basic test by hacking the parameters a bit – i.e. making the decay run every 5 seconds and hack the grace period to be 15mins 😂 

I was able to verify that at least in my test with a single user, it did start decaying once the grace period was hit and keep decaying by the set amount each time the job ran, and was capped at the default sigma (albeit right now it still _tries_ to decays and gets capped, whereas I could move this logic to the initial selection query). 

Additionally verified via psql that it stores the last_game_finished_at correctly and the schema after migration is as expected.

**Comment by mcoot** [comment]
Ok, I've also been able to test that the scheduling based on time from the .env works correctly, as does changing the grace period.

I think this is ready for review however important that we settle on what the right default values should be before we do (note: the feature is still _off_ by default, but if one chooses to enable it, we should decide the default configuration).

Right now it is configured with:
* 1 day default grace period (i.e. if you play every day, you will never have sigma decay applied)
* +0.05 sigma on each decay. This means that if you started from a sigma of say, 2, it would take more than 120 days to decay back to default.

**Comment by mcoot** [comment]
Significant updates:
* The config for sigma decay is now per-category and may be set via the bot rather than needing .env config
* New commands: `!showcategory` (shows details about the category incl. sigma decay info), `!setcategorysigmadecay` to configure the sigma decay.
* An additional configurable variable has been added, which is the maximum proportion of the default sigma a player's sigma may decay to (also configurable via the bot)
* Note there is no longer an env var to disable sigma decay; this is largely unnecessary since the default decay applied for a new category is 0
* Unfortunately I've had to make the db query in the running job worse to support the per-category configuration as I need to compare time intervals based on the value of a column. This is entirely possible in Postgres and in SQLite, but they use completely different syntax for it, so I do the filtering in app code but left a TODO so we can come back to it if we get rid of SQLite

Example of using the commands:

image

File: discord_bots/tasks.py
Comment (StephenBartos) [inline discord_bots/tasks.py:616]: Should be renamed to `sigma_decay_task` to match the style of the existing tasks

File: discord_bots/tasks.py
Comment (StephenBartos) [inline discord_bots/tasks.py:631]: amazing pythonic solution, love how clean your implementation is!

File: discord_bots/cogs/categories.py
Comment (StephenBartos) [inline discord_bots/cogs/categories.py:249]: This can also raise a `sqlalchemy.orm.exc.MultipleResultsFound` exception, so either catch that as well and log it, or use `.first()` instead and check if the return result is `None` -- while `.first()` would ignore the case where multiple categories exist with the same name (which is possible), at least it doesn't break the functionality of the bot (for the most part). I don't really mind what you pick either way, since this is an edge case and the best solution would be to restrict categories from having the same name, which is outside the scope of this PR.

File: discord_bots/tasks.py
Comment (StephenBartos) [inline discord_bots/tasks.py:630]: nit: you might as well filter out categories `where category.sigma_decay_amount != 0`

File: discord_bots/models.py
Comment (StephenBartos) [inline discord_bots/models.py:112]: I'm still not sure if a grace period is useful or just added complexity. Is there a scenario you can think of where you'd want this set to anything other than 0?

**Review by StephenBartos — APPROVED** [review]
Awesome work! I don't see anything blocking this from being merged, aside from a few minor comments and the decision on the default sigma decay (which can be something very small to start, like ~0.02 (1 year to fully decay with default sigma) or ~0.04 (6 months) -- this way people can get acclimated to it without the decay having too large of an effect.
The work you have done here also easily sets up support for mu decay, which would only require it's own set of variables (and then we could move both into a separate model). The update mechanism is solid, thank you for the great code!

File: discord_bots/models.py
Comment (mcoot) [reply discord_bots/models.py:112]: Not certain, but in honesty my initial thought was to set it to 1, which would mean that you never get decayed if you play every day. It doesn't make a whole lot of difference from a TrueSkill perspective, but from a psychological one it may be helpful?

File: discord_bots/cogs/categories.py
Comment (mcoot) [reply discord_bots/cogs/categories.py:249]: This is an issue in all of the category commands at present I'm pretty sure. IMO we should probably add a uniqueness constraint on the name in the db. But I agree that ought to be a separate PR. I can set it to first() for now, but it won't _work_ any better with duplicate names it just won't throw any more.

**Comment by mcoot** [comment]
Re default sigma decay, I'd be happy to set it to 0.02 if we think turning it on by default is a good idea.

**Comment by mcoot** [comment]
Updated with some changes from Discord and GH comments.

File: discord_bots/models.py
Comment (StephenBartos) [reply discord_bots/models.py:112]: If you configure the job to run at 11:59 pm every night, then you wouldn't get decayed if you play that day. I do think it's nice to have the option as an extra dial to tune, but I'm still not entirely convinced it's necessary. 
A scenario where it would be nice, is if you want a very high sigma decay amount but a long grace period: say after 7 days of inactivity you begin decaying players heavily. I'm leaning more towards removing it, since I don't see an immediate usecase but my opinion isn't very strong

Comment (StephenBartos) [reply discord_bots/models.py:112]: On second thought, I think we should keep it. It's a useful parameter

**Comment by StephenBartos** [comment]
> Re default sigma decay, I'd be happy to set it to 0.02 if we think turning it on by default is a good idea.

If the default sigma is `25 / 3`, then for it to take a full year to decay it would need to be (more precisely) `0.023`.
We should figure out whether or not it should be turned on by default with this value (1 year to fully decay at default sigma). I personally lean more towards enabling it by default for any new categories, but the mechanism needs to be explained clearly in the .env.example as well as in the docs

**Comment by mcoot** [comment]
> I personally lean more towards enabling it by default for any new categories, but the mechanism needs to be explained clearly in the .env.example as well as in the docs

Note that as it is currently, it wouldn't enable it only for _new_ categories, but for existing ones too. We can have the DB default be 0 but set it to another value on new category creation though if we'd prefer.

**Comment by mcoot** [comment]
Ok, I've updated it with the following:
* The DB default is now 0 sigma decay, which means existing categories won't have this applied after db upgrade
* But on category _creation_ we now set the value to `default_sigma / 365` so by default it would decay from 0 sigma to default over a year

If this is in a good state would be keen to get it in. We should decide if we want to do this _before_ merging the cog refactor and resolve the conflicts there, or merge that first and resolve the conflicts here. (My guess is the former is going to be a bit easier, but that also gets me off the hook so up to others' views xD)

File: discord_bots/utils.py
Comment (StephenBartos) [inline discord_bots/utils.py:1033]: any reason in particular why this is a function instead of being inlined?

Comment (mcoot) [reply discord_bots/utils.py:1033]: Probably doesn't need to be, to be honest the main reason I ended up moving it out was that the categories cog doesn't directly depends on `config` at all at present

Comment (StephenBartos) [reply discord_bots/utils.py:1033]: No worries, it's not a big deal

**Pull request merged**
